metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,643 @@
+
+        dataset_get_endpoint: str = RegularAPIResolver.API_ENDPOINTS[
+            Constant.DATASET_GET
+        ]
+        # Replace place holders
+        dataset_get_endpoint = dataset_get_endpoint.format(
+            POWERBI_BASE_URL=DataResolverBase.BASE_URL,
+            WORKSPACE_ID=workspace_id,
+            DATASET_ID=dataset_id,
+        )
+        # Hit PowerBi
+        logger.info(f"Request to dataset URL={dataset_get_endpoint}")
+        response = requests.get(
+            dataset_get_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+        # Check if we got response from PowerBi
+        response.raise_for_status()
+        response_dict = response.json()
+        logger.debug("datasets = {}".format(response_dict))
+        # PowerBi Always return the webURL, in-case if it is None then setting complete webURL to None instead of
+        # None/details
+        return new_powerbi_dataset(workspace_id, response_dict)
+
+    def get_groups_endpoint(self) -> str:
+        return DataResolverBase.BASE_URL
+
+    def get_dashboards_endpoint(self, workspace: Workspace) -> str:
+        dashboards_endpoint: str = RegularAPIResolver.API_ENDPOINTS[
+            Constant.DASHBOARD_LIST
+        ]
+        # Replace place holders
+        return dashboards_endpoint.format(
+            POWERBI_BASE_URL=DataResolverBase.BASE_URL, WORKSPACE_ID=workspace.id
+        )
+
+    def get_reports_endpoint(self, workspace: Workspace) -> str:
+        reports_endpoint: str = self.API_ENDPOINTS[Constant.REPORT_LIST]
+        return reports_endpoint.format(
+            POWERBI_BASE_URL=DataResolverBase.BASE_URL, WORKSPACE_ID=workspace.id
+        )
+
+    def get_tiles_endpoint(self, workspace: Workspace, dashboard_id: str) -> str:
+        tiles_endpoint: str = self.API_ENDPOINTS[Constant.TILE_LIST]
+        # Replace place holders
+        return tiles_endpoint.format(
+            POWERBI_BASE_URL=DataResolverBase.BASE_URL,
+            WORKSPACE_ID=workspace.id,
+            DASHBOARD_ID=dashboard_id,
+        )
+
+    def _get_pages_by_report(self, workspace: Workspace, report_id: str) -> List[Page]:
+        pages_endpoint: str = RegularAPIResolver.API_ENDPOINTS[Constant.PAGE_BY_REPORT]
+        # Replace place holders
+        pages_endpoint = pages_endpoint.format(
+            POWERBI_BASE_URL=DataResolverBase.BASE_URL,
+            WORKSPACE_ID=workspace.id,
+            REPORT_ID=report_id,
+        )
+        # Hit PowerBi
+        logger.info(f"Request to pages URL={pages_endpoint}")
+        response = requests.get(
+            pages_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        # Check if we got response from PowerBi
+        response.raise_for_status()
+        response_dict = response.json()
+        return [
+            Page(
+                id="{}.{}".format(report_id, raw_instance["name"].replace(" ", "_")),
```

Comment:
fstring preferred- >
`id=f"{report_id}.{raw_instance["name"]".replace(" ", "_")

Reply:
I converted other places to fstring. Keeping this as is because it contains double quotes to access `name`. It might impact readability if I change it to fstring.